File: packages/cli/src/templates/_shared/lib/base/logger.py
"""Unified logging for all hooks and libraries.

Provides a single logging interface that replaces:
- log_hook_error() from hook_utils.py (error-only, plain text)
- debug.py from cc-native (per-context, plain text)
- eprint() for diagnostic output (stderr-only, no persistence)

Log format: JSONL (one JSON object per line)
Log location: _output/hook-log.jsonl (global, all sessions)
Filter by session using the "sid" field.

Environment variables:
- HOOK_LOG_DISABLE=1: Disable all file logging
- HOOK_LOG_LEVEL=warn: Minimum level to log (default: debug)
- HOOK_ERROR_LOG_DISABLE=1: Legacy alias for HOOK_LOG_DISABLE

Never raises — all errors silently swallowed.
No buffering — each call is one open+write+close.
Stdlib only — json, os, sys, datetime, pathlib.
"""

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, Optional

_LEVELS = {"debug": 0, "info": 1, "warn": 2, "error": 3}

_MAX_LOG_LINES = 10_000  # Max lines in global log before pruning

# Module-level context path cache.
# Set once per hook process via set_context_path() or auto-resolved on first use.
_cached_context_path: Optional[Path] = None
_context_resolved: bool = False

# Module-level session ID cache.
# Set once per hook process via set_session_id().
_cached_session_id: Optional[str] = None


def set_session_id(session_id: Optional[str]) -> None:
    """Set the session ID for this process. All subsequent log calls include it.

    Args:
        session_id: Claude Code session ID (e.g., "a1b2c3d4")
                    or None to clear.
    """
    global _cached_session_id
    _cached_session_id = session_id


def set_context_path(path: Optional[Path]) -> None:
    """Set the context path for this process. All subsequent log calls use it.

    Call this once in your hook after resolving the context:
        from lib.base.logger import set_context_path
        set_context_path(get_context_dir(context_id, project_root))

    Args:
        path: Path to context folder (e.g., _output/contexts/<context-id>/)
              or None to force global-only logging.
    """
    global _cached_context_path, _context_resolved
    _cached_context_path = path
    _context_resolved = True


def _auto_resolve_context_path() -> Optional[Path]:
    """Try to auto-resolve context path from session_id. Called once per process.

    Uses the context store to look up which context owns this session,
    then returns its directory path. Falls back to None (global log).
    """
    global _cached_context_path, _context_resolved
    _context_resolved = True  # Don't retry on failure

    try:
        from ..context.context_store import get_context_by_session_id
        from .constants import get_context_dir

        # Hook input isn't available here, but we can check if a recent
        # context dir exists by scanning _output/contexts/ for one that
        # has a state.json with a matching session
        # This is too expensive for a logger. Instead, rely on set_context_path().
    except Exception:
        pass

    return None


def _get_context_path() -> Optional[Path]:
    """Get the cached context path, auto-resolving on first call."""
    global _context_resolved
    if not _context_resolved:
        _auto_resolve_context_path()
    return _cached_context_path


def _get_min_level() -> int:
    """Get minimum log level from environment."""
    env = os.environ.get("HOOK_LOG_LEVEL", "debug").lower()
    return _LEVELS.get(env, 0)


def _is_disabled() -> bool:
    """Check if file logging is disabled."""
    if os.environ.get("HOOK_LOG_DISABLE") == "1":
        return True
    if os.environ.get("HOOK_ERROR_LOG_DISABLE") == "1":
        return True
    return False


def _get_project_root() -> Path:
    """Get project root from environment or cwd."""
    env_dir = os.environ.get("CLAUDE_PROJECT_DIR", "")
    return Path(env_dir) if env_dir else Path.cwd()


def hook_log(
    level: str,
    hook_name: str,
    message: str,
    *,
    component: str = "",
    data: Any = None,
    traceback_str: str = "",
    stderr: bool = True,
) -> None:
    """Write a structured log entry to the global hook log.

    All entries go to _output/hook-log.jsonl. Use the "sid" field
    (set via set_session_id) to filter by session.

    Args:
        level: "debug" | "info" | "warn" | "error"
        hook_name: Hook or module name (e.g., "session_end")
        message: Log message
        component: Sub-component (e.g., "git", "parse")
        data: Optional structured data (must be JSON-serializable)
        traceback_str: Optional traceback string
        stderr: Also write to stderr (default: True)
    """
    try:
        level_lower = level.lower()
        level_num = _LEVELS.get(level_lower, 0)

        # Write to stderr if requested
        if stderr:
            prefix = f"[{hook_name}]"
            if component:
                prefix = f"[{hook_name}:{component}]"
            print(f"{prefix} {message}", file=sys.stderr)
            if traceback_str:
                print(traceback_str, file=sys.stderr)

        # Check if file logging is enabled
        if _is_disabled():
            return

        # Check minimum level
        if level_num < _get_min_level():
            return

        # Build JSONL entry
        ts = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
        entry: Dict[str, Any] = {
            "ts": ts,
            "level": level_lower,
            "hook": hook_name,
            "msg": message,
        }
        if _cached_session_id:
            entry["sid"] = _cached_session_id
        if component:
            entry["component"] = component
        if data is not None:
            try:
                json.dumps(data, default=str)  # Validate serializable
                entry["data"] = data
            except (TypeError, ValueError):
                entry["data"] = str(data)
        if traceback_str:
            entry["tb"] = traceback_str.rstrip()

        line = json.dumps(entry, ensure_ascii=True, default=str) + "\n"

        # Always write to global log
        project_root = _get_project_root()
        log_path = project_root / "_output" / "hook-log.jsonl"

        log_path.parent.mkdir(parents=True, exist_ok=True)

        # Line-count guard: prune to last _MAX_LOG_LINES
        if log_path.exists():
            try:
                with open(log_path, "r", encoding="utf-8", errors="replace") as f:
                    existing = f.readlines()
                if len(existing) > _MAX_LOG_LINES:
                    with open(log_path, "w", encoding="utf-8") as f:
                        f.writelines(existing[-_MAX_LOG_LINES:])
            except OSError:
                pass

        with open(log_path, "a", encoding="utf-8") as f:
            f.write(line)

    except Exception:
        pass  # Never crash


def log_debug(hook_name: str, message: str, **kwargs: Any) -> None:
    """Log a debug-level message."""
    hook_log("debug", hook_name, message, **kwargs)


def log_info(hook_name: str, message: str, **kwargs: Any) -> None:
    """Log an info-level message."""
    hook_log("info", hook_name, message, **kwargs)


def log_warn(hook_name: str, message: str, **kwargs: Any) -> None:
    """Log a warn-level message."""
    hook_log("warn", hook_name, message, **kwargs)


def log_error(hook_name: str, message: str, **kwargs: Any) -> None:
    """Log an error-level message."""
    hook_log("error", hook_name, message, **kwargs)


def log_diagnostic(
    hook_name: str,
    phase: str,
    summary: str,
    *,
    inputs: Any = None,
    decision: Any = None,
    reasoning: Any = None,
    component: str = "diag",
    data: Any = None,
) -> None:
    """Log a structured diagnostic entry at a hook decision point.

    Emits a debug-level JSONL entry with tagged, filterable data.
    Use at key decision points: receive (what came in), decide (what was chosen),
    result (what happened).

    Args:
        hook_name: Hook or module name (e.g., "session_start")
        phase: Decision phase — "receive", "decide", or "result"
        summary: One-line description (e.g., "source=clear, session=a1b2c3d4")
        inputs: Input data relevant to this phase
        decision: The decision made (for "decide" phase)
        reasoning: Why this decision was made
        component: Log component tag (default: "diag")
        data: Extra data to merge into the structured entry
    """
    diag_data: Dict[str, Any] = {"phase": phase}
    if inputs is not None:
        diag_data["inputs"] = inputs
    if decision is not None:
        diag_data["decision"] = decision
    if reasoning is not None:
        diag_data["reasoning"] = reasoning
    if data is not None and isinstance(data, dict):
        diag_data.update(data)
    hook_log(
        "debug",
        hook_name,
        f"[DIAG:{phase}] {summary}",
        component=component,
        data=diag_data,
    )


def log_hook_error(
    hook_name: str,
    error: Exception,
    hook_event: str = "unknown",
    traceback_str: str = "",
) -> None:
    """Backward-compatible wrapper matching the old hook_utils.log_hook_error signature.

    Delegates to hook_log("error", ...) with the same behavior:
    - Message capped at 200 chars, newlines stripped
    - Never raises

    Args:
        hook_name: Name of the hook
        error: The exception that occurred
        hook_event: Hook event type (e.g., "PreToolUse")
        traceback_str: Optional formatted traceback
    """
    msg = str(error).replace("\n", " ").replace("\r", "")[:200]
    err_type = type(error).__name__
    hook_log(
        "error",
        hook_name,
        f"[{hook_event}] {err_type}: {msg}",
        traceback_str=traceback_str,
        stderr=True,
    )

File: packages/cli/src/templates/_shared/lib-ts/base/logger.ts
/**
 * Unified logging for all hooks and libraries.
 *
 * Log format: JSONL (one JSON object per line)
 * Log location: _output/hook-log.jsonl (global, all sessions)
 * Filter by session using the "sid" field.
 *
 * stderr is OPT-IN: convenience functions (logDebug, logInfo, logWarn, logError)
 * write to file only by default. To also write to stderr (visible to Claude Code
 * as "hook error"), pass { stderr: true } or use logBlocking().
 * logHookError() always writes to stderr (unhandled errors must be visible).
 *
 * Environment variables:
 * - HOOK_LOG_DISABLE=1: Disable all file logging
 * - HOOK_LOG_LEVEL=warn: Minimum level to log (default: debug)
 * - HOOK_ERROR_LOG_DISABLE=1: Legacy alias for HOOK_LOG_DISABLE
 *
 * Never throws. No buffering. Stdlib only.
 * See SPEC.md §3
 */

import * as fs from "node:fs";
import * as path from "node:path";

const LEVELS: Record<string, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
};

const MAX_LOG_LINES = 10_000; // Max lines in global log before pruning

// Module-level session ID cache
let _cachedSessionId: string | null = null;

// Module-level context path cache (kept for external callers)
let _cachedContextPath: string | null = null;
let _contextResolved = false;

/**
 * Set the session ID for this process. All subsequent log calls include it.
 */
export function setSessionId(sessionId: string | null): void {
  _cachedSessionId = sessionId;
}

/**
 * Set the context path for this process. Kept for external callers.
 */
export function setContextPath(contextPath: string | null): void {
  _cachedContextPath = contextPath;
  _contextResolved = true;
}

export function getContextPath(): string | null {
  if (!_contextResolved) {
    _contextResolved = true; // Don't retry
  }
  return _cachedContextPath;
}

function getMinLevel(): number {
  const env = (process.env.HOOK_LOG_LEVEL ?? "debug").toLowerCase();
  return LEVELS[env] ?? 0;
}

function isDisabled(): boolean {
  return (
    process.env.HOOK_LOG_DISABLE === "1" ||
    process.env.HOOK_ERROR_LOG_DISABLE === "1" ||
    process.env.CCNATIVE_DEBUG_DISABLE === "1"
  );
}

function getProjectRoot(): string {
  return process.env.CLAUDE_PROJECT_DIR || process.cwd();
}

/**
 * Write a structured log entry to the global hook log.
 *
 * All entries go to _output/hook-log.jsonl. Use the "sid" field
 * (set via setSessionId) to filter by session.
 */
export function hookLog(
  level: string,
  hookName: string,
  message: string,
  opts?: {
    component?: string;
    data?: any;
    traceback_str?: string;
    stderr?: boolean;
  },
): void {
  try {
    const levelLower = level.toLowerCase();
    const levelNum = LEVELS[levelLower] ?? 0;
    const component = opts?.component ?? "";
    const tracebackStr = opts?.traceback_str ?? "";
    const stderrEnabled = opts?.stderr === true;

    // Write to stderr
    if (stderrEnabled) {
      const prefix = component
        ? `[${hookName}:${component}]`
        : `[${hookName}]`;
      process.stderr.write(`${prefix} ${message}\n`);
      if (tracebackStr) {
        process.stderr.write(tracebackStr + "\n");
      }
    }

    // Check if file logging is enabled
    if (isDisabled()) return;

    // Check minimum level
    if (levelNum < getMinLevel()) return;

    // Build JSONL entry
    const now = new Date();
    const ts = now.toISOString().replace("Z", "").slice(0, 23);
    const entry: Record<string, any> = {
      ts,
      level: levelLower,
      hook: hookName,
      msg: message,
    };
    if (_cachedSessionId) entry.sid = _cachedSessionId;
    if (component) entry.component = component;
    if (opts?.data !== undefined && opts.data !== null) {
      try {
        JSON.stringify(opts.data);
        entry.data = opts.data;
      } catch {
        entry.data = String(opts.data);
      }
    }
    if (tracebackStr) entry.tb = tracebackStr.trimEnd();

    const line = JSON.stringify(entry) + "\n";

    // Always write to global log
    const logPath = path.join(getProjectRoot(), "_output", "hook-log.jsonl");

    // Ensure directory exists
    const dir = path.dirname(logPath);
    fs.mkdirSync(dir, { recursive: true });

    // Line-count guard: prune to last MAX_LOG_LINES
    try {
      if (fs.existsSync(logPath)) {
        const content = fs.readFileSync(logPath, "utf-8");
        const lines = content.split(/\r?\n/);
        if (lines.length > MAX_LOG_LINES) {
          fs.writeFileSync(
            logPath,
            lines.slice(lines.length - MAX_LOG_LINES).join("\n"),
            "utf-8",
          );
        }
      }
    } catch {
      // ignore
    }

    fs.appendFileSync(logPath, line, "utf-8");
  } catch {
    // Never crash
  }
}

export function logDebug(hookName: string, message: string, opts?: Record<string, any>): void {
  hookLog("debug", hookName, message, opts);
}

export function logInfo(hookName: string, message: string, opts?: Record<string, any>): void {
  hookLog("info", hookName, message, opts);
}

export function logWarn(hookName: string, message: string, opts?: Record<string, any>): void {
  hookLog("warn", hookName, message, opts);
}

export function logError(hookName: string, message: string, opts?: Record<string, any>): void {
  hookLog("error", hookName, message, opts);
}

/**
 * Log an error that SHOULD be visible to user/model via stderr.
 * Use for real problems needing attention, not routine diagnostics.
 */
export function logBlocking(hookName: string, message: string, opts?: Record<string, any>): void {
  hookLog("error", hookName, message, { ...opts, stderr: true });
}

/**
 * Log a structured diagnostic entry at a hook decision point.
 * See SPEC.md §3.8
 */
export function logDiagnostic(
  hookName: string,
  phase: string,
  summary: string,
  opts?: {
    inputs?: any;
    decision?: any;
    reasoning?: any;
    component?: string;
    data?: any;
  },
): void {
  const diagData: Record<string, any> = { phase };
  if (opts?.inputs !== undefined) diagData.inputs = opts.inputs;
  if (opts?.decision !== undefined) diagData.decision = opts.decision;
  if (opts?.reasoning !== undefined) diagData.reasoning = opts.reasoning;
  if (opts?.data && typeof opts.data === "object") {
    Object.assign(diagData, opts.data);
  }
  hookLog("debug", hookName, `[DIAG:${phase}] ${summary}`, {
    component: opts?.component ?? "diag",
    data: diagData,
  });
}

/**
 * Backward-compatible wrapper matching old hook_utils.log_hook_error signature.
 * See SPEC.md §3.6
 */
export function logHookError(
  hookName: string,
  error: Error | string,
  hookEvent = "unknown",
  tracebackStr = "",
): void {
  const errStr = typeof error === "string" ? error : String(error);
  const msg = errStr.replace(/[\n\r]/g, " ").slice(0, 200);
  const errType =
    typeof error === "object" && error !== null
      ? error.constructor.name
      : "Error";
  hookLog("error", hookName, `[${hookEvent}] ${errType}: ${msg}`, {
    traceback_str: tracebackStr,
    stderr: true,
  });
}
